Src/Base/AMReX_VisMF.cpp

```diff
@@ -2471,6 +2518,828 @@ VisMF::WriteAsync (const FabArray<FArrayBox>& mf, const std::string& mf_name)
+                  "WriteAsync: unsupported Real size");
+
+#ifdef AMREX_USE_MPI
+    int thread_support = -1;
+    MPI_Query_thread(&thread_support);
+    AMREX_ALWAYS_ASSERT(thread_support == MPI_THREAD_MULTIPLE);
+#endif
+
+    const DistributionMapping& dm = mf.DistributionMap();
+
+    int myproc = ParallelDescriptor::MyProc();
+    int nprocs = ParallelDescriptor::NProcs();
+
+    RealDescriptor const& whichRD = []() -> RealDescriptor const& {
+        switch (FArrayBox::getFormat())
+        {
+        case FABio::FAB_NATIVE:
+            return FPC::NativeRealDescriptor();
+        case FABio::FAB_NATIVE_32:
+            return FPC::Native32RealDescriptor();
+        case FABio::FAB_IEEE_32:
+            return FPC::Ieee32NormalRealDescriptor();
+        default:
+            return FPC::NativeRealDescriptor();
+        }
+    }();
+    bool doConvert = whichRD != FPC::NativeRealDescriptor();
+
+    VisMF::Header hdr(mf, VisMF::NFiles, VisMF::Header::Version_v1, false);
+
+    auto myinfo = StaticWriteInfo(myproc);
+    int ifile = std::get<0>(myinfo);   // file #
+    int ispot = std::get<1>(myinfo);   // spot #
+
+    constexpr int sizeof_int64_over_real = sizeof(int64_t) / sizeof(Real);
+    const int n_local_fabs = mf.local_size();
+    const int n_global_fabs = mf.size();
+    const int ncomp = mf.nComp();
+    const long n_fab_reals = 2*ncomp;
+    const long n_fab_int64 = 1;
+    const long n_fab_nums = n_fab_reals*sizeof_int64_over_real + n_fab_int64;
+    const long n_local_nums = n_fab_nums * n_local_fabs + 1;
+    Vector<int64_t> localdata(n_local_nums);
+
+    RunOn runon;
+#ifdef AMREX_USE_GPU
+    if (Gpu::inLaunchRegion() &&
+        mf.arena() == The_Arena() ||
+        mf.arena() == The_Device_Arena() ||
+        mf.arena() == The_Pinned_Arena())
+    {
+        runon = RunOn::Device;
+    } else {
+        runon = RunOn::Host;
+    }
+#else
+    runon = RunOn::Host;
+#endif
+
+
+
+    int64_t total_bytes = 0;
+    auto pld = (char*)(&(localdata[1]));
+    const FABio& fio = FArrayBox::getFABio();
+    for (MFIter mfi(mf); mfi.isValid(); ++mfi)
+    {
+        std::memcpy(pld, &total_bytes, sizeof(int64_t));
+        pld += sizeof(int64_t);
+
+        const FArrayBox& fab = mf[mfi];
+
+        std::stringstream hss;
+        fio.write_header(hss, fab, ncomp);
+        total_bytes += static_cast<std::streamoff>(hss.tellp());
+        total_bytes += fab.size() * whichRD.numBytes();
+
+        // compute min and max
+        const Box& bx = mfi.validbox();
+
+        for (int icomp = 0; icomp < ncomp; ++icomp) {
+            if (runon == RunOn::Host)
+            {
+                Real cmin = fab.min<RunOn::Host>(bx,icomp);
+                Real cmax = fab.max<RunOn::Host>(bx,icomp);
+                std::memcpy(pld, &cmin, sizeof(Real));
+                pld += sizeof(Real);
+                std::memcpy(pld, &cmax, sizeof(Real));
+                pld += sizeof(Real);
```

Comment:
Maybe we can move this and the similar code in `else if` out of the if block to remove duplication.